File: src/javacode/RSACipher.java
package javacode;

import org.jetbrains.annotations.NotNull;

import javax.crypto.Cipher;
import java.nio.charset.StandardCharsets;
import java.security.*;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;

public class RSACipher {
    private final byte[] publicKeyBytes;
    private byte[] privateKeyBytes = null;

    public RSACipher() {
        KeyPair keyPair = generateKeyPair();
        publicKeyBytes = keyPair.getPublic().getEncoded();
        privateKeyBytes = keyPair.getPrivate().getEncoded();
    }
    public RSACipher(@NotNull KeyPair keyPair) {
        publicKeyBytes = keyPair.getPublic().getEncoded();
        privateKeyBytes = keyPair.getPrivate().getEncoded();
    }
    public RSACipher(@NotNull PublicKey publicKey, @NotNull PrivateKey privateKey) {
        publicKeyBytes = publicKey.getEncoded();
        privateKeyBytes = privateKey.getEncoded();
    }
    public RSACipher(@NotNull String publicKey, @NotNull String privateKey) {
        publicKeyBytes = Base64.getDecoder().decode(publicKey.getBytes(StandardCharsets.UTF_16LE));
        privateKeyBytes = Base64.getDecoder().decode(privateKey.getBytes(StandardCharsets.UTF_16LE));
    }
    public RSACipher(@NotNull String publicKey) {
        publicKeyBytes = Base64.getDecoder().decode(publicKey.getBytes(StandardCharsets.UTF_16LE));
    }
    public RSACipher(byte[] publicKey, byte[] privateKey) {
        publicKeyBytes = publicKey;
        privateKeyBytes = privateKey;
    }
    public RSACipher(byte[] publicKey) {
        publicKeyBytes = publicKey;
    }

    @NotNull
    public static String getKeyString(@NotNull Key key) {
        byte[] bytes = Base64.getEncoder().encodeToString(key.getEncoded()).getBytes();
        return new String(bytes, StandardCharsets.UTF_16LE);
    }
    @NotNull
    public static String getKeyString(byte[] keyBytes) {
        byte[] bytes = Base64.getEncoder().encodeToString(keyBytes).getBytes();
        return new String(bytes, StandardCharsets.UTF_16LE);
    }

    @NotNull
    public static KeyPair generateKeyPair() {
        KeyPair keyPair = null;
        try {
            KeyPairGenerator keyPairGen = KeyPairGenerator.getInstance("RSA");
            keyPairGen.initialize(2048);
            keyPair = keyPairGen.generateKeyPair();
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }
        assert keyPair != null;
        return keyPair;
    }

    @NotNull
    public String getPublicKey() {
        return getKeyString(publicKeyBytes);
    }
    @NotNull
    public String getPrivateKey() {
        return getKeyString(privateKeyBytes);
    }

    public String encrypt(String string) {
        if (string == null || string.isEmpty()) {
            return null;
        }

        try {
            PublicKey publicKey = KeyFactory.getInstance("RSA").generatePublic(new X509EncodedKeySpec(publicKeyBytes));

            Cipher cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
            cipher.init(Cipher.ENCRYPT_MODE, publicKey);

            byte[] encryptedBytes = cipher.doFinal(string.getBytes(StandardCharsets.UTF_8));
            String encryptedString = Base64.getEncoder().encodeToString(encryptedBytes);
            return new String(encryptedString.getBytes(), StandardCharsets.UTF_16LE);
        } catch (Exception e){
            e.printStackTrace();
            return null;
        }
    }
    public static String encrypt(String string, @NotNull PublicKey publicKey) {
        if (string == null || string.isEmpty()) {
            return null;
        }

        try {
            Cipher cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
            cipher.init(Cipher.ENCRYPT_MODE, publicKey);

            byte[] encryptedBytes = cipher.doFinal(string.getBytes(StandardCharsets.UTF_8));
            String encryptedString = Base64.getEncoder().encodeToString(encryptedBytes);
            return new String(encryptedString.getBytes(), StandardCharsets.UTF_16LE);
        } catch (Exception e){
            e.printStackTrace();
            return null;
        }
    }
    public static String encrypt(String string, @NotNull String publicKeyString) {
        if (string == null || string.isEmpty()) {
            return null;
        }

        try {
            byte[] publicKeyBytes = Base64.getDecoder().decode(publicKeyString.getBytes(StandardCharsets.UTF_16LE));
            PublicKey publicKey = KeyFactory.getInstance("RSA").generatePublic(new X509EncodedKeySpec(publicKeyBytes));

            Cipher cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
            cipher.init(Cipher.ENCRYPT_MODE, publicKey);

            byte[] encryptedBytes = cipher.doFinal(string.getBytes(StandardCharsets.UTF_8));
            String encryptedString = Base64.getEncoder().encodeToString(encryptedBytes);
            return new String(encryptedString.getBytes(), StandardCharsets.UTF_16LE);
        } catch (Exception e){
            e.printStackTrace();
            return null;
        }
    }

    public String decrypt(String string) {
        if (string == null || string.isEmpty()) {
            return null;
        }

        try {
            PrivateKey privateKey = KeyFactory.getInstance("RSA").generatePrivate(new PKCS8EncodedKeySpec(privateKeyBytes));

            Cipher cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
            cipher.init(Cipher.DECRYPT_MODE, privateKey);


            byte[] encryptedBytes = Base64.getDecoder().decode(string.getBytes(StandardCharsets.UTF_16LE));
            byte[] decryptedBytes = cipher.doFinal(encryptedBytes);

            return new String(decryptedBytes, StandardCharsets.UTF_8);
        } catch (Exception e){
            e.printStackTrace();
            return null;
        }
    }
    public static String decrypt(String string, @NotNull PrivateKey privateKey) {
        if (string == null || string.isEmpty()) {
            return null;
        }

        try {
            Cipher cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
            cipher.init(Cipher.DECRYPT_MODE, privateKey);

            byte[] encryptedBytes = Base64.getDecoder().decode(string.getBytes(StandardCharsets.UTF_16LE));
            byte[] decryptedBytes = cipher.doFinal(encryptedBytes);

            return new String(decryptedBytes, StandardCharsets.UTF_8);
        } catch (Exception e){
            e.printStackTrace();
            return null;
        }
    }
    public static String decrypt(String string, @NotNull String privateKeyString) {
        if (string == null || string.isEmpty()) {
            return null;
        }

        try {
            byte[] privateKeyBytes = Base64.getDecoder().decode(privateKeyString.getBytes(StandardCharsets.UTF_16LE));
            PrivateKey privateKey = KeyFactory.getInstance("RSA").generatePrivate(new PKCS8EncodedKeySpec(privateKeyBytes));

            Cipher cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
            cipher.init(Cipher.DECRYPT_MODE, privateKey);

            byte[] encryptedBytes = Base64.getDecoder().decode(string.getBytes(StandardCharsets.UTF_16LE));
            byte[] decryptedBytes = cipher.doFinal(encryptedBytes);

            return new String(decryptedBytes, StandardCharsets.UTF_8);
        } catch (Exception e){
            e.printStackTrace();
            return null;
        }
    }
}

File: src/kotlin/RSACipher.kt
package kotlin

import java.nio.charset.StandardCharsets.UTF_16LE
import java.security.NoSuchAlgorithmException
import java.security.Key
import java.security.KeyFactory
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import java.util.Base64
import javax.crypto.Cipher

object RSACipher {
    val Key.string: String
        get() {
            val bytes = Base64.getEncoder().encodeToString(this.encoded).toByteArray()
            return String(bytes, UTF_16LE)
        }

    val generateKeyPair: KeyPair
        get() {
            var keyPair: KeyPair? = null
            try {
                val keyPairGen = KeyPairGenerator.getInstance("RSA")
                keyPairGen.initialize(2048)
                keyPair = keyPairGen.generateKeyPair()
            } catch (e: NoSuchAlgorithmException) {
                e.printStackTrace()
            }
            assert(keyPair != null)
            return keyPair!!
        }

    fun encrypt(string: String?, publicKeyString: String): String? {
        if (string.isNullOrEmpty()) {
            return null
        }

        try {
            val publicKeyBytes = Base64.getDecoder().decode(publicKeyString.toByteArray(UTF_16LE))
            val publicKey = KeyFactory.getInstance("RSA").generatePublic(X509EncodedKeySpec(publicKeyBytes))

            val cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding")
            cipher.init(Cipher.ENCRYPT_MODE, publicKey)

            val encryptedBytes = cipher.doFinal(string.toByteArray(UTF_16LE))
            val encryptedString = Base64.getEncoder().encodeToString(encryptedBytes)
            return String(encryptedString.toByteArray(), UTF_16LE)
        } catch (e: Exception) {
            e.printStackTrace()
            return null
        }
    }

    fun decrypt(string: String?, privateKeyString: String): String? {
        if (string.isNullOrEmpty()) {
            return null
        }

        try {
            val privateKeyBytes =
                Base64.getDecoder().decode(privateKeyString.toByteArray(UTF_16LE))
            val privateKey = KeyFactory.getInstance("RSA").generatePrivate(PKCS8EncodedKeySpec(privateKeyBytes))

            val cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding")
            cipher.init(Cipher.DECRYPT_MODE, privateKey)

            val encryptedBytes = Base64.getDecoder().decode(string.toByteArray(UTF_16LE))
            val decryptedBytes = cipher.doFinal(encryptedBytes)

            return String(decryptedBytes, UTF_16LE)
        } catch (e: Exception) {
            e.printStackTrace()
            return null
        }
    }
}
